### PR TITLE
[self-profiler] Make the profiler faster/more efficient

### PR DESCRIPTION
**Pull request by wesleywiser**
Related to #58372

r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Once #58309 lands, we should do a perf run to measure the overhead. I suspect that emitting JSON instead of binary data is still more expensive than we'd like it to be.

**Comment by wesleywiser** [comment]
Probably, I haven't looked into doing binary serialization yet. Is there any existing infrastructure in the compiler for that or should I roll my own?

**Comment by wesleywiser** [comment]
(FYI, in case you were remembering the performance of serialization as it was at All-Hands, that was without buffering the writes. I added a `BufWriter` around the file and it significantly improved the performance.)

**Comment by michaelwoerister** [comment]
There is rustc's [libserialize](https://github.com/rust-lang/rust/tree/master/src/libserialize) which works OK for metadata and incremental caches.

**Comment by michaelwoerister** [comment]
I don't know how easy or hard it is to use `libserialize` outside of the compiler though. Let's check how the JSON version does first.

**Comment by wesleywiser** [comment]
I think it's just a matter of `#[derive(RustcDecodable)]` but I'm not sure.

**Comment by michaelwoerister** [comment]
> I think it's just a matter of `#[derive(RustcDecodable)]` but I'm not sure.

Yes, I think that's true as long as both the compiler and importing tool are compiled with the same compiler version.

Given that there's just a couple of data structures that need to be serialized, doing something handwritten might be a viable option though. `libserialize` has some overhead because it uses leb128 for encoding integers. Since we care more about speed than about file size `libserialize` might not be a good trade off.

**Comment by michaelwoerister** [comment]
Can you switch this to use a `parking_lot::Mutex`? Then we can do a first perf run to see how be the overhead is now.

**Comment by wesleywiser** [comment]
Sure, I can do that tonight. Are you trying to measure the record overhead or the save results overhead? 

I'll need to push a commit to enable whichever option you want to record by default since the entire feature is disabled by default.

**Comment by michaelwoerister** [comment]
I think we should do both.

**Comment by wesleywiser** [comment]
Ok. Do you want to do seperate perf runs or just one with the option turned on by default?

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -1401,6 +1401,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "run the self profiler"),
     profile_json: bool = (false, parse_bool, [UNTRACKED],
         "output a json file with profiler results"),
+    dump_profiler_events: bool = (false, parse_bool, [UNTRACKED],
```
Comment (wesleywiser) [inline src/librustc/session/config.rs]: This option isn't used and the other profiler related options should be cleaned up.

**Comment by michaelwoerister** [comment]
@rust-timer build 695f06aa4019c5cfa5c884a12f0d35270a65afae

**Comment by wesleywiser** [comment]
That's... actually not as terrible as I thought it was going to be 😄

**Comment by wesleywiser** [comment]
@michaelwoerister I remember at All Hands, @eddyb had the idea to use an MPSC channel to push events to a profiler thread. Perhaps it's worth trying that and seeing that the performance looks like doing it that way? 

I imagine we can lower the max RSS and wall clock time quite a bit by eagerly writing events to disk.

**Comment by michaelwoerister** [comment]
That's one option. The other would be to collect events per thread into a memory mapped file and let the OS take care of writing stuff to disk in the background.

The background thread approach has a few disadvantages:
- You need an additional CPU core in order for profiling not to take away resources from the code being profiled.
- Doing all of this cross-thread communication might cause quite a bit of overhead and trying to optimize this away might make the implementation rather complicated.

Per-thread event recording also has disadvantages:
- A session would produce a number of files instead of just one. This makes the program doing the post processing more complicated. But actually it should not be *that* complicated.
- It might have to rely on `Drop::drop` being reliably run for thread-local storage -- or at least a simple implementation would do that. There might be more complicated ways to implement this that don't need to rely on this.

But it also has advantages:
- It might not really show up in max-rss because the OS can page out the parts of memory that have already been written (they are never touched again). I'm not sure about this though.
- There is no or almost no cross-thread communication happening, allowing for things to be rather efficient.

Overall I think per-thread collection has more potential.

Maybe someone in @rust-lang/wg-compiler-performance has a good idea on how to keep the overhead of event tracking and persistence as low as possible?

**Comment by nagisa** [comment]
A per-thread approach would result in inconsistent order between what has really happened and what got profiled. Timers needn’t be sequentially consistent between threads (as proved by a recent Windows bug).

All that being said, I *still* think that per-thread approach is better than the others. However, all threads should be writing into the same buffer, reserving space in the buffer in a lock-free manner. Naive sketch of an algorithm:

```rust
static INDEX: AtomicUsize = 0;
static BUFFER: &[u8; however long it needs to be] = /* memory mapped? */;

fn emit(event: Event) {
    let message = event.serialize();
    let my_base = INDEX.fetch_add(message.len(), SeqCst); // Something that also resizes the buffer, if necessary (possibly not an issue for memory mapped files? Make it a ring buffer that something consumes somewhere?)
    BUFFER[my_base..(my_base + message.len)].copy_from_slice(message);    
}

// or:

fn emit(event: Event) {
    let message = event.serialize();
    let my_base = INDEX.fetch_add(message.len(), SeqCst);
    pwrite(logfile, message.ptr(), message.len(), my_base);
}
```

This would ensure the least amount of overhead without losing sequential consistency between the events.

**Comment by michaelwoerister** [comment]
So, looking at a callgrind profile of doing a clean check build of `regex`, it looks like:
 - 30% of the instructions are spent formatting profiling events to JSON
 - 4.5% of instructions are spent in event recording. 

Interestingly the biggest portion of `Profiler::record` execution is spent in generating a SipHash. That should be an easy fix `:)`

The bigger problem is emitting events as JSON. I have some ideas for how to encode them in binary without generating too much overhead. I'll post them to the tracking issue (https://github.com/rust-lang/rust/issues/58372) once I've sorted them some more...

That's looks like a very smart concurrent allocation strategy, @nagisa!

**Comment by michaelwoerister** [comment]
So, to give an update here: I think we can merge this
- after switching to `FxHashMap` and
- making sure that it doesn't break tools unexpectedly (because it does change some of the session interfaces)

My expectation is still that we won't use JSON because it's too costly. But that can be done in a future PR.

**Comment by wesleywiser** [comment]
> * making sure that it doesn't break tools unexpectedly (because it does change some of the session interfaces)

@michaelwoerister would you like me to revert the change to the session interfaces for now? We could pursue that later if we're still seeing missing time at the beginning. I think there are some other PRs to the compiler interfaces in flight at the moment like #58250 so perhaps it's a good idea to defer that particular change anyway.

**Comment by michaelwoerister** [comment]
Yes, let's do that. We can still keep the changes that place the profiler in an `Arc` so that codegen/LLVM has access to it.

**Comment by wesleywiser** [comment]
Yay CI finally passed! I removed the changes to the session interfaces and switched the `HashMap` to an `FxHashMap`. This should be ready for review.

**Comment by michaelwoerister** [comment]
Looks good, thanks @wesleywiser!

@bors r+

**Comment by wesleywiser** [comment]
Rebased

@bors r+

**Comment by michaelwoerister** [comment]
@bors r+

(let's not make it look like you self-approved the PR `:)`)

**Comment by wesleywiser** [comment]
Oh, sorry. Not sure what the etiquette is there :)

**Comment by michaelwoerister** [comment]
You usually just write `@ bors r=name-of-whoever-approved-the-PR-initially`. Like this for example:
https://github.com/rust-lang/rust/pull/58605#issuecomment-467782099

**Comment by wesleywiser** [comment]
Rebased

@bors r=michaelwoerister

**Comment by wesleywiser** [comment]
Rebased

@bors r=michaelwoerister

**Comment by jens1o** [comment]
What is the perf-graph looking like today?

**Comment by Mark-Simulacrum** [comment]
This won't affect perf.r-l.o since it isn't run with self-profile enabled.

**Comment by michaelwoerister** [comment]
I still expect it to be rather slow. Here are some thoughts on how to make it faster: #58372

**Pull request merged**